src/core/manager.c

```diff
@@ -126,6 +127,107 @@ static int manager_run_environment_generators(Manager *m);
 static int manager_run_generators(Manager *m);
 static void manager_vacuum(Manager *m);
 
+int manager_spawn_workers(Manager *m) {
+        char socket_fd_number[DECIMAL_STR_MAX(int) + 1];
+        _cleanup_free_ char *log_level = NULL;
+        static int n_workers_cache = -2;
+        const char *scope_path;
+        size_t n_workers;
+        int r;
+
+        assert(m);
+
+        if (FLAGS_SET(m->test_run_flags, MANAGER_TEST_RUN_MINIMAL))
+                return 0;
+
+        if (n_workers_cache < -1) {
+                const char *e = secure_getenv("SYSTEMD_WORKERS_POOL_SIZE");
+                if (e) {
+                        r = safe_atoi(e, &n_workers_cache);
+                        if (r < 0) {
+                                log_debug("Invalid value in $SYSTEMD_WORKERS_POOL_SIZE, ignoring: %s", e);
+                                n_workers_cache = -1;
+                        }
+                } else
+                        n_workers_cache = -1;
+        }
+
+        if (n_workers_cache == 0)
+                return 0;
+
+        scope_path = strjoina(m->cgroup_root, "/" SPECIAL_INIT_WORKERS_SCOPE);
+
+        r = cg_is_empty(SYSTEMD_CGROUP_CONTROLLER, scope_path);
+        if (r <= 0)
+                return r;
+
+        /* At boot we are going to fork off a lot of processes on average, so prepare a bunch. */
+        if (n_workers_cache >= 0)
+                n_workers = n_workers_cache;
+        else if (MANAGER_IS_SYSTEM(m) && IN_SET(manager_state(m), MANAGER_STARTING, MANAGER_INITIALIZING))
+                n_workers = 25;
+        else
+                n_workers = 5;
+
+        r = log_level_to_string_alloc(log_get_max_level(), &log_level);
+        if (r < 0)
+                return log_error_errno(r, "Failed to convert log level to string: %m");
+
+        xsprintf(socket_fd_number, "%i", m->executors_pool_socket[1]);
+
+        r = fd_cloexec(m->executors_pool_socket[1], false);
+        if (r < 0)
+                return log_error_errno(errno, "Failed to set O_CLOEXEC on serialization fd: %m");
+
+        for (size_t i = 0; i < n_workers; ++i) {
+                pid_t pid;
+
+                /* The executor binary is pinned, to avoid compatibility problems during upgrades. */
+                r = posix_spawn_wrapper(FORMAT_PROC_FD_PATH(m->executor_fd),
+                                STRV_MAKE(FORMAT_PROC_FD_PATH(m->executor_fd),
+                                        "--manager-socket", socket_fd_number,
+                                        "--log-level", log_level,
+                                        "--log-target", log_target_to_string(manager_get_executor_log_target(m))),
+                                environ,
+                                &pid);
+                if (r < 0)
+                        return log_error_errno(r, "Failed to spawn executor: %m");
+
+                (void) cg_attach(SYSTEMD_CGROUP_CONTROLLER, scope_path, pid);
```

Comment:
hmm, i am pretty sure the executor should also do this btw as first step, beforing processing any requests, so that whatever it does is attributed properly to the right cgroup.

Reply:
and yes, it's fine if both child and parent do this. one will be a nop, but that doesn't matter

Reply:
It's already done in exec_invoke() as one of the first things. I intentionally left it like that, because if something goes wrong in the handshake, I want to be able to zap the whole workers cgroup and start over, so that I am 100% sure it is starting from a clean slate